extensions/cache/deployment/src/test/java/io/quarkus/cache/test/runtime/UniReturnTypeTest.java

```diff
@@ -0,0 +1,188 @@
+package io.quarkus.cache.test.runtime;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotSame;
+import static org.junit.jupiter.api.Assertions.assertSame;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.cache.CacheInvalidate;
+import io.quarkus.cache.CacheInvalidateAll;
+import io.quarkus.cache.CacheResult;
+import io.quarkus.test.QuarkusUnitTest;
+import io.smallrye.mutiny.Uni;
+
+/**
+ * Tests the caching annotations on methods returning {@link Uni}.
+ */
+public class UniReturnTypeTest {
+
+    private static final String CACHE_NAME_1 = "test-cache-1";
+    private static final String CACHE_NAME_2 = "test-cache-2";
+    private static final String KEY_1 = "key-1";
+    private static final String KEY_2 = "key-2";
+
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest().withApplicationRoot((jar) -> jar.addClass(CachedService.class));
+
+    @Inject
+    CachedService cachedService;
+
+    @Test
+    void testCacheResult() {
+        // STEP 1
+        // Action: a method annotated with @CacheResult and returning a Uni is called.
+        // Expected effect: the method is not invoked, as Uni is lazy.
```

Comment:
I think that I understand the reasoning but I find this rather counter-intuitive. As the method is annotated with `@CacheResult` I would expect the method to be executed but the result of the invocation to be cached. But I'm no reactive expert so it might be perfectly OK.

@cescoffier @jponge WDYT?

Reply:
Everything (invocation and caching of the result) happens at subscription time. This PR doesn't change that part of the extension BTW. I refactored `CacheResultInterceptor` a bit to get rid of runtime lambdas, nothing more.

Reply:
Yep, I'm not saying that it's a change or anything but this is the first time I see it... and it's not documented anyway so ;-)

Reply:
It was changed recently though with https://github.com/quarkusio/quarkus/pull/22271.

Reply:
> and it's not documented anyway so ;-)

It will be shortly, I'm on it 😄

Reply:
I would naively expect that the method is invoked, the result is cached and a next invocation would return a new `Uni` for the cached result. But like I said - I'm no expert in this area at all.

Reply:
That always puzzled me. 
With:

```java
@CahcedResult
Uni<Person> getPerson() { ... }
```

What is cached? The `Uni`? The emitted `Person`?

If you cache the `Uni`, resubscribing will re-execute the logic, so probably not what we want to achieve.
Caching the person makes more sense to me, but I'm happy to discuss it.

Reply:
The original `Uni` is never cached by the extension. What is cached is an [UnresolvedUniValue](https://github.com/quarkusio/quarkus/blob/main/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/UnresolvedUniValue.java) at first which is then replaced with the actual item emitted by the original `Uni` once it's been resolved. After that, when the cached value is accessed, a `Uni.createFrom().item(theResolvedValue)` is returned.

Reply:
> In other words, the method body is not called until someone subscribes to the `Uni` which is returned by the interceptor, right? That probably makes sense.

Exactly.

> My naive expectation was that the method is called and once someone subscribes to the returned `Uni` a `Uni.createFrom().item(theResolvedValue)` is returned. That's very similar but I assume the current approach makes more sense from caching POV.

The method was invoked before the subscription before but it caused a bug with resteasy-reactive (blocked thread) which was [fixed](https://github.com/quarkusio/quarkus/pull/22271) by @stuartwdouglas.